pennylane/interfaces/jacobian_products.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Defines classes that take the vjps, jvps, and jacobians of circuits.
+
+"""
+import abc
+from functools import partial
+from typing import Tuple, Callable, Optional
+
+import pennylane as qml
+from pennylane.tape import QuantumScript
+from pennylane.typing import ResultBatch, TensorLike
+
+Batch = Tuple[QuantumScript]
+
+
+class JacobianProductCalculator(abc.ABC):
+    """Provides methods for calculating the jvp and vjps for tapes and tangents/ cotangents."""
+
+    @abc.abstractmethod
+    def execute_and_compute_jvp(
+        self, tapes: Batch, tangents: TensorLike
```

Comment:
Should the type hint for `tangents` be a `Sequence` of `TensorLike` or `Sequence[TensorLike]`? :thinking: